src/java/org/apache/cassandra/schema/TableParams.java

```diff
@@ -490,6 +490,11 @@ public Builder extensions(Map<String, ByteBuffer> val)
 
     public static class Serializer implements MetadataSerializer<TableParams>
     {
+        private final String keyspace;
```

Comment:
I dont think this is a good idea, because then you have `new Serializer...." everywhere. These serializers are meant to be effectively singletons, rather put `keyspace` as parameter of the method instead

Reply:
we're pulling in `keyspace` from the context when we shouldn't be (the call to `CompressionParams.defaultParams(keyspace)` needs to be kept in the context).

i'm not sure yet how to do this…

Reply:
@Claudenw  @michaelsembwever 

I tried to fix it here https://github.com/Claudenw/cassandra/pull/12/files

It moved a lot of changes back. They are not necessary at all. Just because we expected keyspace in build method as a param made them more verbose than necessary.

I think it is also not necessary to call it every time, if one reads the context where that code is used, sometimes it is not necessary to deal with compression params again etc ...

Reply:
I don't think that using it just upon deserialization is enough. We want to set this up correctly in whole database context when creating table metadata etc ...

Reply:
I dont think we should change that build() method and deal with that in it. It should be as it was. Putting there null is also strange ... Let's @michaelsembwever be an arbiter who decides what is better. Currently we just make more changes that necessary. Putting keyspace into build() is just weird imho.

Reply:
we could actually do this https://github.com/Claudenw/cassandra/pull/12/commits/b88cecc712b07c497ccbe55552bfe09aa2b5e91a

so there is not need to have any parameters in build() method at all because we can set that on `params` when building the metadata.

Reply:
@Claudenw btw you are right that this does not prevent a user to set default custom compressor for system tables. This still needs to be covered. I will think about that ... But in general I think that this is a right direction.

Reply:
hah I realized that ....


    cqlsh> ALTER TABLE system.local WITH compression = {'class': 'ZstdCompressor'};
    Unauthorized: Error from server: code=2100 [Unauthorized] message="system keyspace is not user-modifiable."

So we are already covered on that front. But if somebody wanted to set that programmatically, we can do this:

https://github.com/Claudenw/cassandra/pull/12/commits/ea26c2e8cf684563dc88d909dd02ba9c2669b62b

Reply:
@Claudenw after my changes, that "partiallyBuild" is not necessary, I have remove that  https://github.com/Claudenw/cassandra/pull/12/commits/20c790a255f89da08506b728a8170d36d6ef6865

so, we are not having partially anywhere, we are not putting keyspace anywhere, we are checking that system is always on default and ser/de is not touched at all.

Reply:
I think that his whole conversation is not relevant anymore by CASSANDRA-19592